### ISSUE TITLE
[SOLVED] Unable to start swank server in NixOS

### ISSUE THREAD

**Issue by rollschild** [kovisoft/slimv#120]
Hi,

I'm a Common Lisp beginner. I installed Slimv and sbcl. Now when I tried to evaluate a simple hello world lisp file using `,e`, vim said `Starting SWANK server...` then after a few seconds returned error `SWANK server is not running`. Does anyone know how to fix this? Thank you so much!

**Comment by rollschild** [comment]
Hi,

So I copied and pasted 
```
! xterm -e sbcl --load /nix/store/ccw0alq4m7zzx5rldrr280bkycnf2h1l-vimplugin-slimv-2021-08-24/slime/start-swank.lisp &
```
into NeoVim command after `:` and hit Enter and nothing happened.

I typed `:! ls` and it showed me the correct output, but didn't open up a new shell. It just printed the output in the default Vim output area, at the bottom of Vim, as shown in this picture:
![IMG_20220302_205549](https://user-images.githubusercontent.com/6456628/156484234-9832e86f-72c7-4f04-b7e0-3f4f5a8162aa.jpg)

I typed in `:! xterm`, it _did_ open up a new window, which was a shell. So I think this part works correctly.
![IMG_20220302_205628](https://user-images.githubusercontent.com/6456628/156484432-3698be91-9aa8-4df2-aad2-341b3e069fb5.jpg)

If I omitted the `&` in the original command and typed it in NeoVim, it successfully opened up a new shell with SWANK server successfully connected. However, I don't know how to give control back to NeoVim to test whether I can actually connect to the server so I wasn't able to test.

![IMG_20220302_205839](https://user-images.githubusercontent.com/6456628/156484556-6157431b-b22b-40ce-b0d3-d2342aa5ae57.jpg)

As you suggested, as a workaround, I could launch this command in terminal to open up a new shell and then open a lisp file separately in NeoVim. This worked perfectly well. I was able to connect to the server and evaluate.

So, to summarize, I guess the issue is somehow NeoVim couldn't detect or work with command like `:! xterm some_command &`, which, is supposed to run this command in the background?

**Comment by rollschild** [comment]
I posted a question in NeoVim's discourse forum: https://neovim.discourse.group/t/what-is-the-correct-recommended-way-to-run-a-shell-command-in-neovim-in-the-background/2130?u=rollschild. I will post here if there are any replies to that question.

**Comment by kovisoft** [comment]
Thank you for the feedback. Are you absolutely sure that the xterm command with `&` at the end really did nothing? `&` means the given command should be executed in the background and immediately return control to the calling process. Can you check what processes are running on your system? Maybe xterm is running but you cannot see it because it is running in the background. Well, that's the main purpose of `&` at the end of the swank command: the swank server should be started in the background, but control should be returned to vim.

**Comment by rollschild** [comment]
I figured it out.

I was right that the `xterm` command was never fired by that `:! xterm .... &` command in NeoVim. I was running a `ps a | grep xterm` command in the same tmux session and found nothing.

I had to use NeoVim's own [job control](https://neovim.io/doc/user/job_control.html) mechanism to sort of "spawn" a process. What I added to `/ect/configuration.nix` was this:
```
            let g:slimv_swank_cmd = "call jobstart('xterm -e & sbcl --load ${pkgs.vimPlugins.slimv}/slime/start-swank.lisp')"
```

To non-NixOS users, it would be something like 
```
let g:slimv_swank_cmd = "call jobstart('xterm -e & sbcl --load /usr/share/common-lisp/source/slime/start-swank.lisp')"
```

Basically, I just changed the original `let g:slimv_swank_cmd = '! xterm -e sbcl --load /usr/share/common-lisp/source/slime/start-swank.lisp &'` to a more NeoVim way.

You can consider this as closed now. Thanks for your help!

**Comment by kovisoft** [comment]
Great! Thank you for the feedback.
